### PR TITLE
[Blazor] Startup API cleanups

### PR DESCRIPTION
**Pull request by javiercn**
Implements changes from API review:
* Move AddRenderMode to a separate class in the infrastructure namespace, (no extension method). 
* Call that method from AddWebAssemblyRenderMode and AddServerRenderMode.
* Change WebAssemblyOptions for callback.

### CODE REVIEW

File: src/Components/WebAssembly/Server/src/WebAssemblyRazorComponentsBuilderExtensions.cs [hunk 1]
```diff
@@ -32,25 +30,21 @@ public static IRazorComponentsBuilder AddWebAssemblyComponents(this IRazorCompon
...
-            _options = options.Value;
-        }
-
         public override IEnumerable<RouteEndpointBuilder> GetEndpointBuilders(IComponentRenderMode renderMode, IApplicationBuilder applicationBuilder)
         {
             if (renderMode is not WebAssemblyRenderModeWithOptions wasmWithOptions)
             {
+                if (renderMode is WebAssemblyRenderMode)
+                {
+                    throw new InvalidOperationException("Invalid render mode. Use AddWebAssemblyComponents(Action<WebAssemblyComponentsEndpointOptions>) to configure the WebAssembly render mode.");
```
Comment (halter73) [inline src/Components/WebAssembly/Server/src/WebAssemblyRazorComponentsBuilderExtensions.cs:41]: Would it make sense to make a similar check in `CircuitEndpointProvider`? We could add an internal type in `AddServerRenderMode` similar to `WebAssemblyRenderModeWithOptions` even if it doesn't configure any options yet.

If we don't do it now, we wouldn't be able to add a similar check in the future without a breaking change if we decide to add a `ServerComponentsEndpointOptions`. It feels weird to allow `ComponentEndpointConventionBuilderHelper.AddRenderMode(builder, new MyServerRenderMode())` to work when `new MyWebAssemblyRenderMode()` cannot.

**Review by halter73 — APPROVED** [review]

**Review by SteveSandersonMS — APPROVED** [review]
Agreed with @halter73's point but besides that looks great!

File: src/Components/WebAssembly/Server/src/WebAssemblyRazorComponentsBuilderExtensions.cs (hunk 1 above)
Comment (javiercn) [reply src/Components/WebAssembly/Server/src/WebAssemblyRazorComponentsBuilderExtensions.cs]: done

**Comment by mkArtakMSFT** [comment]
Merging as all the CI check failures are expected due to API changes.

**Pull request merged**